pkg/ensurance/analyzer/sort/evict-sort.go

```diff
@@ -0,0 +1,182 @@
+package sort
+
+import (
+	"sort"
+
+	v1 "k8s.io/api/core/v1"
+
+	"github.com/gocrane/crane/pkg/ensurance/executor"
+	"github.com/gocrane/crane/pkg/utils"
+)
+
+// EvictRankFunc sorts the pods in eviction order
+type EvictRankFunc func(pods []executor.EvictPod)
+
+var evictSortFunc = map[string]func(p1, p2 executor.EvictPod) int{
+	"ExtResourceEnable": evictUseExtResource,
+	"ClassAndPriority":  evictClassAndPriority,
+	"ExtResourceUsage":  evictExtCpuUsed2Limit,
+	"CpuUsage":          evictCpuUsage,
+	"RunningTime":       evictStartTime,
+}
+
+func EvictRankFuncConstruct(customize []string) EvictRankFunc {
+	var evictRankFunc EvictRankFunc
+	if len(customize) == 0 {
+		cmp := []cmpEvictFunc{}
+		for _, sortFunc := range customize {
+			if f, ok := evictSortFunc[sortFunc]; ok {
+				cmp = append(cmp,f)
+			}
+			evictRankFunc = evictOrderedBy(cmp...).Sort
+			return evictRankFunc
+		}
+	}
+	evictRankFunc = evictRankCpuPressure
+
+	return evictRankFunc
+}
+
+func evictRankCpuPressure(pods []executor.EvictPod) {
+	evictOrderedBy(evictUseExtResource, evictClassAndPriority, evictExtCpuUsed2Limit, evictCpuUsage, evictStartTime).Sort(pods)
+}
+
+// evictStartTime compares pods by pod's start time
+func evictStartTime(p1, p2 executor.EvictPod) int {
+	t1 := p1.StartTime
+	t2 := p2.StartTime
+
+	if t1 == nil && t2 == nil {
+		return 0
+	}
+	if t1 == nil {
+		// maybe t1 is just starting
+		return 1
+	}
+	if t2 == nil {
+		return -1
+	}
+
+	if t1.Before(t2) {
+		return 1
+	} else if t1.Equal(t2) {
+		return 0
+	}
+	return -1
+}
+
+// evictClassAndPriority compares pods by pod's ClassAndPriority
+func evictClassAndPriority(p1, p2 executor.EvictPod) int {
+	return int(executor.CompareClassAndPriority(p1.ClassAndPriority, p2.ClassAndPriority))
+}
+
+// evictUseExtResource compares pod by whether use ext resource
+func evictUseExtResource(p1, p2 executor.EvictPod) int {
+	use1 := utils.TransBool2Uint(p1.UseExtResource)
+	use2 := utils.TransBool2Uint(p2.UseExtResource)
+
+	if use1 == use2 {
+		return 0
+	}
+	if use1 < use2 {
+		return -1
+	}
+	return 1
+}
+
+// evictExtCpuUsed2Limit compares the partition extcpu usage of extcpu limit
+func evictExtCpuUsed2Limit(p1, p2 executor.EvictPod) int {
+	// if both pod don't use ext resource, then return
+	if p1.UseExtResource == false && p2.UseExtResource == false {
```

Comment:
if one pod is use extResource but another is not, we return zero is suitable?

Reply:
It will return in the previous sort func: extCpuBeUsed